src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs

```diff
@@ -1,351 +1,1805 @@
 		}
 
-		protected override bool? IsHeightConstrainedInner(NativeView requester)
+		//------------------------------------------------------------------------
+		//
+		//  Method:   MeasureCellsGroup
+		//
+		//  Synopsis: Measure one group of cells
+		//
+		//------------------------------------------------------------------------
+		void MeasureCellsGroup(
+			int cellsHead, //cell group number
+			int cellCount, //elements in the cell
+			XFLOAT rowSpacing,
+			XFLOAT columnSpacing,
+			bool ignoreColumnDesiredSize,
+			bool forceRowToInfinity,
+			ref CellCacheStackVector cellCacheVector)
 		{
-			if (requester == null)
+
+			SpanStoreStackVector spanStore = new SpanStoreStackVector();
+
+			if (cellsHead >= cellCount)
 			{
-				// The Grid itself called invalidate
-				return this.IsHeightConstrainedSimple();
+				//goto Cleanup;
+				return;
 			}
 
-			var row = GetRow(requester);
-			var rowSpan = GetRowSpan(requester);
-			rowSpan = Math.Max(rowSpan, 1);
-			for (int i = row; i < row + rowSpan; i++)
+			do
 			{
-				if (i >= RowDefinitions.Count)
+				CellCache cell = cellCacheVector[cellsHead];
+				UIElement pChild = cell.m_child;
+
+				MeasureCell(pChild, cell.m_rowHeightTypes, cell.m_columnWidthTypes, forceRowToInfinity, rowSpacing,
+					columnSpacing);
+				//If a span exists, add to span store for delayed processing. processing is done when
+				//all the desired sizes for a given definition index and span value are known.
+
+				if (!ignoreColumnDesiredSize)
 				{
-					return this.IsHeightConstrainedSimple();
+					Xuint columnSpan = GetColumnSpanAdjusted(pChild);
+					//pChild.EnsureLayoutStorage();
+					if (columnSpan == 1)
+					{
+						DefinitionBase pChildColumn = GetColumnNoRef(pChild);
+						pChildColumn.UpdateEffectiveMinSize((XFLOAT)pChild.DesiredSize.Width);
+					}
+					else
+					{
+						RegisterSpan(
+							spanStore,
+							GetColumnIndex(pChild),
+							columnSpan,
+							pChild.DesiredSize.Width,
+							true /* isColumnDefinition */);
+					}
 				}
-				GridLength height = RowDefinitions[i].Height;
-				if (height.IsAuto || height.IsStar)
+
+				if (!forceRowToInfinity)
 				{
-					//At least one of the rows spanned by the requester is auto/star-sized, so height may change
-					return false;
+					Xuint rowSpan = GetRowSpanAdjusted(pChild);
+					//pChild.EnsureLayoutStorage();
+					if (rowSpan == 1)
+					{
+						DefinitionBase pChildRow = GetRowNoRef(pChild);
+						pChildRow.UpdateEffectiveMinSize((XFLOAT)pChild.DesiredSize.Height);
+					}
+					else
+					{
+						RegisterSpan(
+							spanStore,
+							GetRowIndex(pChild),
+							rowSpan,
+							pChild.DesiredSize.Height,
+							false /* isColumnDefinition */);
+					}
 				}
-			}
-			for (int i = row; i < row + rowSpan; i++)
+
+				cellsHead = cellCacheVector[cellsHead].m_next;
+
+			} while (cellsHead < cellCount);
+
+			//Go through the spanned rows/columns allocating sizes.
+			//for (auto & entry : spanStore.m_vector)
```

Comment:
Probably should remove all those as well

Reply:
```suggestion
```